text/0000-relative-kernels.md

```diff
@@ -0,0 +1,203 @@
+
+# Community-level explanation
+[community-level-explanation]: #community-level-explanation
+
+Explain the proposal as if it were already introduced into the Grin ecosystem and you were teaching it to another community member. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Grin community members should *think* about the improvement, and how it should impact the way they interact with Grin. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Grin community members and new Grin community members.
+
+For implementation-oriented RFCs (e.g. for wallet), this section should focus on how wallet contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Grin currently supports three kernel variants -
+
+* Plain
+* Coinbase
+* HeightLocked
+
+These are implemented as kernel "feature" variants -
+
+```
+pub enum KernelFeatures {
+	/// Plain kernel (the default for Grin txs).
+	Plain = 0,
+	/// A coinbase kernel.
+	Coinbase = 1,
+	/// A kernel with an explicit lock height.
+	HeightLocked = 2,
+}
+```
+
+Each kernel variant supports feature specific data -
+
+```
+# Plain
+{    
+  "fee": 8
+}
+
+# Coinbase
+{
+  # empty
+}
+
+# Height Locked
+{
+  "fee": 8,
+  "lock_height": 295800
+}
+```
+
+We propose adding an additional feature variant "Relative Height Locked".
+
+```
+pub enum KernelFeatures {
+	...
+	/// A kernel with a relative lock height.
+	RelativeHeightLocked = 3,
+}
+```
+
+A relative height locked kernel would reference a previous kernel by its excess (commitment)
+along with a relative height (height in blocks between the two kernels).
+
+```
+# Relative Height Locked
+{
+  "fee": 8,
+  "lock_height": 1044,
+  "rel_kernel": "088305235baac64e90daca81b0bad7afbce3a6e49c989572095a892e857e681429"
+}
+```
+
+In the example above the transaction containing this kernel would only be valid 1044 blocks (approx 24 hours) after the referenced kernel.
+
+While we require a full 32 bytes of data to reference a previous kernel we can optimize this when storing the kernel in the MMR. We append kernels to the MMR based on immutable history and this allows us to reference the prior kernel by MMR position. At a given block height the MMR is immutable and the kernel at a given position is deterministic and immutable. Every node sees a consistent view of kernel ordering in their local copy of the kernel MMR.
```

Comment:
We limit the recent_blocks to at most WEEK_HEIGHT, so that a one week window of recent kernels suffices to check validity of relative kernels.